### PR TITLE
Frontend Unit, Integration, & E2E Testing Part 1

### PR DESCRIPTION
**Pull request by abbiehuynh**
## This PR includes:
**Implementing a comprehensive testing strategy using Cypress that includes:**
- Unit testing for individual components, 
- Integration testing to verify component interactions,
- E2E testing to simulate user interactions

**Data test attributes** were implemented to components to provide a stable way to select elements without being affected by changes in the UI, such as CSS class names or element structure.
- This ensures that even if the button’s class or other attributes change, the test can still correctly identify and interact with the button as long as the data-test attribute remains the same.

Based on test results, the sort date function in the Project Provider was updated to improve date sorting functionality. 

## Testing Screenshots:
Renders and tests functionality of Login component
login-testing

Rendering components on homepage
home-testing

Rendering Navigation Bar and testing navigation to correct page when link is clicked
navbar-testing

Rendering Search bar and testing search bar functionality
searchbar-testing

Rendering Filter and Sort headers and buttons; testing button functionality
sortprojectstatus-testing

Renders and tests functionality of buttons on Project Card (favorite, open project, and delete)
projectCard - testing

Renders and tests functionality of Add Project Modal and Form
add-project-testing

Renders and tests functionality of AI Chatbot Modal and Form
ai-testing

### Unit, Integration, & E2E Tests of components to be added in Part 2:
- Project Details Page
- Project Form
- Editing Project Form
- Updating Project Form 
- Favorites Page
- Register & Registration Form
- Logout

### Notes:
- Due to the large amount of changes made, the Frontend Unit, Integration, & E2E testing will be implemented in two parts. 
- Please review the testing files and suggest any improvements on testing approach or suggestions for additional tests.

---
Merging mvp-testing branch into mvp branch

### CODE REVIEW

File: client/src/components/ai/AIChatBot.jsx [hunk 1]
```diff
@@ -46,7 +46,7 @@ const AIChatBot = ({ isOpen, onClose }) => {
     };
 
     return (
-        <Modal show={isOpen} onHide={onClose}>
+        <Modal show={isOpen} onHide={onClose} data-test="ai-chat-modal">
```
Comment (davidmedina-contact) [inline client/src/components/ai/AIChatBot.jsx:49]: It'd be good to add a script to remove these from the code for production deployments as a future improvement

File: client/src/components/home/ProjectProvider.jsx [hunk 2]
```diff
@@ -78,9 +78,9 @@ export const ProjectProvider = ({ children }) => {
                 const typeB = b.project_type || '';
                 return typeA.localeCompare(typeB);
             case 'date': 
-            // sort by updated date descending
-            return new Date(b.updated_at) - new Date(a.updated_at); 
-            
+                const dateA = new Date(a.created_at);
+                const dateB = new Date(b.created_at);
+                return dateB - dateA; // sort by date descending
```
Comment (davidmedina-contact) [inline client/src/components/home/ProjectProvider.jsx:83]: Why is this change needed?

File: cypress.config.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+const { defineConfig } = require('cypress')
+
+module.exports = defineConfig({
+  e2e: {
+    baseUrl: 'http://localhost:5173',
+    setupNodeEvents(on, config) {
+      // implement node event listeners here
+    },
```
Comment (davidmedina-contact) [inline cypress.config.js]: This can be removed

File: cypress/e2e/addProject.cy.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+describe('Add Project Modal Functionality', () => {
+    beforeEach(() => {
+        // login using custom command found in /support/commands.js
+        cy.login('kingliver', 'cats');
+    });
+
+    it('renders add project button', () => {
+        cy.get('[data-test="add-project-btn"]').should('be.visible')
+          .and('contain', 'Add Project');
+      });
```
Comment (davidmedina-contact) [inline cypress/e2e/addProject.cy.js]: I wonder if testing the presence of a button that is not controled by any user interaction or code logic is worth it?

File: cypress/e2e/addProject.cy.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        cy.get('[data-test="add-project-modal"]').should('be.visible');
+        // types the project name into the input
+        cy.get('[data-test="add-project-name-input"]')
+            .type(projectName)
+            .should('have.value', projectName);
+        // submits the form
+        cy.get('[data-test="add-project-form"]').submit();
+        // verifies that the confirmation model appears and is correct
+        cy.get('[data-test="confirmation-modal"]').should('be.visible');
+        cy.get('[data-test="confirmation-modal"]').contains(`Project "${projectName}" has been added successfully!`);
```
Comment (davidmedina-contact) [inline cypress/e2e/addProject.cy.js]: For future improvements, the message strings that you want to test against should be in language/localization strings so you don't have to hard code expected text in tests

File: cypress/e2e/addProject.cy.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+        // // if the user clicks home
+        // // closes the confirmation modal to go to home
+        // cy.get('[data-test="home-btn-confirmation-modal"]').click();
+        // // checks that the user is redirected to the home page
+        // cy.url().should('include', '/home');
+        // // checks if the project is displayed in the project list
+        // cy.get('[data-test="project-list"]').contains(projectName).should('exist');
```
Comment (davidmedina-contact) [inline cypress/e2e/addProject.cy.js]: Why are these commented out?

File: cypress/e2e/addProject.cy.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        
+        // if the user clicks edit project
+        // closes the confirmation modal to go to project details page
+        cy.get('[data-test="edit-project-btn-confirmation-modal"]').click();
+        // checks that the user is redirected to the project details page
+        cy.url().should('include', `/user/4/project/`);
```
Comment (davidmedina-contact) [inline cypress/e2e/addProject.cy.js]: Why is this user `4`? This might make the test finicky and prone to fail as you evolve your app. Find a way to programatically set this route

File: cypress/e2e/ai.cy.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
+// testing suite for ai chatbot component
+// add it.only on line 10 to prevent calling open ai
+
+describe('AI ChatBot Functionality', () => {
+    beforeEach(() => {
+        // login using custom command found in /support/commands.js
+        cy.login('kingliver', 'cats');
```
Comment (davidmedina-contact) [inline cypress/e2e/ai.cy.js:7]: Is this a user that gets created in the db just for testing purposes?

File: cypress/e2e/ai.cy.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+        // opens the modal
+        cy.get('[data-test="ai-btn"]').click();
+        // types a message
+        cy.get('[data-test="ai-input"]').type(userMessage);
+        cy.get('[data-test="ai-submit-btn"]').click();
+        // checks that the message appears in the chat history
+        cy.get('[data-test="ai-chat-history"]').first().should('contain', `You: ${userMessage}`);
+        // checks that the bot's response is visible in the chat history // can not wait on request, had to use timeout
+        cy.get('[data-test="ai-chat-history"]', { timeout: 10000 }).last().should('not.be.empty');
```
Comment (davidmedina-contact) [inline cypress/e2e/ai.cy.js:34]: What is displayed after the timeout?

File: cypress/e2e/addProject.cy.js [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        cy.get('[data-test="confirmation-modal"]').contains(`Project "${projectName}" has been added successfully!`);
+
+        // // if the user clicks home
+        // // closes the confirmation modal to go to home
+        // cy.get('[data-test="home-btn-confirmation-modal"]').click();
+        // // checks that the user is redirected to the home page
+        // cy.url().should('include', '/home');
+        // // checks if the project is displayed in the project list
+        // cy.get('[data-test="project-list"]').contains(projectName).should('exist');
+
```
Comment (tati-guzman) [inline cypress/e2e/addProject.cy.js]: Why is this commented out? The test below it (checking for the project id to be in the project lists) seems similar but given that this checks for the route and update back on the home page, this test would seem valuable as well if that's a functionality you still expect

File: cypress/e2e/home.cy.js [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+  it('renders sort header and buttons', () => {
+    cy.get('[data-test="sort-header"]').should('be.visible');
+    cy.get('[data-test="sort-name"]').should('be.visible');
+    cy.get('[data-test="sort-type"]').should('be.visible');
+    cy.get('[data-test="sort-date"]').should('be.visible');
+  });
+
+  it('renders project list header', () => {
+    cy.get('[data-test="project-header"]').should('be.visible');
+  })
```
Comment (davidmedina-contact) [inline cypress/e2e/home.cy.js:37]: I wonder if these are worth testing if they are also not dependent of user interaction or any code/API call

File: cypress/e2e/ai.cy.js [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    });
```
Comment (tati-guzman) [inline cypress/e2e/ai.cy.js:35]: No specific feedback here other than to tell you this test was fascinating! So impressed with this feature on your project!

File: cypress/e2e/login.cy.js [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    cy.get('[data-test="login-username"]').should('have.value', 'kingliver');
+  })
+
+  it('checks password input', () => {
+    cy.visit('/login');
+
+    cy.get('[data-test="login-password"]').type('cats');
+    cy.get('[data-test="login-password"]').should('have.value', 'cats');
+  })
+});
```
Comment (davidmedina-contact) [inline cypress/e2e/login.cy.js:32]: Same here. Testing that inputs accept text is not something you coded, but rather expected behavior of HTML components. Unless you added some logic that controls how the inputs react to how users type?

File: cypress/e2e/login.cy.js [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    // insert valid username and password
+    cy.get('[data-test="login-username"]').type('kingliver'); 
+    cy.get('[data-test="login-password"]').type('cats');
+
+    // clicks the login button
+    cy.get('[data-test="login-btn"]').click();
+
+    // checks that the user is redirected to the home page
+    cy.url().should('include', '/home');
+    // chcecks that 'Your Projects' header is shown on home page
```
Comment (tati-guzman) [inline cypress/e2e/login.cy.js]: Just wanted to point out the tiniest spelling mistake in this comment!

File: cypress/e2e/navbar.cy.js [hunk 15]
```diff
@@ -0,0 +1,37 @@
+// testing suite for navbar component
+
+describe('Navigation Bar', () => {
+    beforeEach(() => {
+      // login using custom command found in /support/commands.js
+      cy.login('kingliver', 'cats');
```
Comment (davidmedina-contact) [inline cypress/e2e/navbar.cy.js:6]: Might be worth to look into a way to avoid having to log in every time with something like Cypress sessions: https://www.cypress.io/blog/authenticate-faster-in-tests-cy-session-command

File: cypress/e2e/login.cy.js [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+    cy.visit('/login');
+
+    // insert invalid username and password and click button
+    cy.get('[data-test="login-username"]').type('kingliverr'); 
+    cy.get('[data-test="login-password"]').type('cats');
+    cy.get('[data-test="login-btn"]').click();
+
+    // checks that error message is displayed
+    cy.contains('Invalid username or password').should('be.visible');
+  })
```
Comment (tati-guzman) [inline cypress/e2e/login.cy.js:62]: Love that you also checked for invalid credentials!!

File: cypress/e2e/projectCard.cy.js [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+        cy.get('[data-test^="project-card-"]').first().find('[data-test="favorite-btn"]')
+        .should('contain.text', 'Click to favorite project');
+    });
+
+    it('handles favorite button toggle', () => {
+        // finds first project card and clicks favorite button
+        cy.get('[data-test^="project-card-"]').first().find('[data-test="favorite-btn"]').click();
+        // checks if favorite icon is red
+        cy.get('[data-test^="project-card-"]').first().find('[data-test="favorite-btn"] svg')
+            .should('have.attr', 'style').and('include', 'color: red');
```
Comment (davidmedina-contact) [inline cypress/e2e/projectCard.cy.js]: Were you still not abler to contol this with a CSS class?

File: cypress/e2e/projectCard.cy.js [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+    it.only('renders the favorite button and hidden span', () => {
+        cy.get('[data-test^="project-card-"]').each(($card) => {
+            cy.wrap($card).find('[data-test="favorite-btn"]').should('be.visible');
+        });
+
+        cy.get('[data-test^="project-card-"]').first().find('[data-test="favorite-btn"]')
+        .should('contain.text', 'Click to favorite project');
+    });
+
+    it('handles favorite button toggle', () => {
```
Comment (davidmedina-contact) [inline cypress/e2e/projectCard.cy.js:20]: What about testing the case of removing a favorite?

File: cypress/e2e/projectCard.cy.js [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+        });
+    });
+
+    it('handles delete button functionality', () => {
+         // finds first project card and clicks delete button
+        cy.get('[data-test^="project-card-"]').first().find('[data-test="delete-btn"]').click();
+        // automatically confirm any confirmation dialog
+        cy.on('window:confirm', () => true);
+        // verifies that the project card is no longer present
+        cy.get('[data-test^="project-card-"]').should('have.length.lessThan', 10);
```
Comment (davidmedina-contact) [inline cypress/e2e/projectCard.cy.js]: Is there a better way to test this? Like confirming the card id is not present in the URL? Relying on the length of cards is finicky

File: cypress/e2e/register.cy.js
Comment (davidmedina-contact) [inline cypress/e2e/register.cy.js:1]: I think it's better to not include testing files for pages that have no tests _yet_

File: client/src/components/auth/Login.jsx [hunk 20]
```diff
@@ -45,18 +45,19 @@ const Login = () => {
                 setError(error.response.data.message || 'An error occured. Please try again.');
                 console.error('Error response:', error.response.data)
             }
-            setError('An error occurred. Please try again.');
+            setError('Invalid username or password');
             console.error('Error:', error);
         }
     };
```
Comment (tati-guzman) [inline client/src/components/auth/Login.jsx:51]: Is this a user displayed message or console logged? I only asked because I noticed the test expects the error message to be visible but I remembered this being set for console.error and couldn't find a user version. If it doesn't exist yet, I would recommend presenting to the user as well.

File: cypress/e2e/searchBar.cy.js [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+      it('allows user to type in search bar', () => {
+        const searchProject = 'Silly Goose';
+       
+        // assigns search input element to an alias - recieved error that required less chaining
+        cy.get('[data-test="search-bar-input"]').as('searchInput');
+        // types project name into input
+        cy.get('@searchInput').type(searchProject);
+        // requery the element to check it's still attached to the DOM
+        cy.get('@searchInput').should('have.value', searchProject);
+      });
```
Comment (davidmedina-contact) [inline cypress/e2e/searchBar.cy.js:23]: Same here. I don't think there's a lot of value testing this

File: cypress/e2e/searchBar.cy.js [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+      it('displays the project after searching', () => {
+       const searchProject = 'Silly Goose';
+
+       cy.get('[data-test="search-bar-input"]').type(searchProject);
+       cy.wait(500); // waits for the project list to update
+
+       // checks the project card appears and for the specific project card
+       cy.get('[data-test="project-list"]').should('be.visible');
+       cy.get('[data-test^=project-card-]').should('have.length.greaterThan', 0);
+      })
```
Comment (davidmedina-contact) [inline cypress/e2e/searchBar.cy.js:34]: Are these projects created in the db for testing purposes? If not, tests might be finicky because they depend on a predetermined db structure

File: cypress/e2e/sortProjectStatus.cy.js [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+
+      it('filters projects when "In Progress" is clicked', () => {
+        cy.get('[data-test="filter-inprogress"]').click();
+        cy.get('[data-test="project-list"]').should('contain', 'In Progress');
+      })
+
+      it('filters projects when "Completed" is clicked', () => {
+        cy.get('[data-test="filter-completed"]').click();
+        cy.get('[data-test="project-list"]').should('contain', 'Completed');
+      })
```
Comment (davidmedina-contact) [inline cypress/e2e/sortProjectStatus.cy.js:38]: Same here (about predetermined db data)

File: cypress/e2e/projectCard.cy.js [hunk 24]
```diff
@@ -0,0 +1,75 @@
+// testing suite for project card component 
+// testing functionality of favorite, open project, and delete button
+// added it.only to line 11, to prevent accidentally deleting projects
```
Comment (tati-guzman) [inline cypress/e2e/projectCard.cy.js:3]: I'm unfamiliar with it.only. Since you test for the delete button functionality as a separate test below, why was this needed as a precaution for the favorites button?

File: cypress/fixtures/example.json
Comment (davidmedina-contact) [inline cypress/fixtures/example.json:1]: Don't commit this file

**Review by tati-guzman — APPROVED** [review]
Awesome job on your testing, Abbie! You were so thorough, detailed, and organized! I left a few comments and questions that are mostly for my own curiosity and I don't think impede these changes being merged so I am setting my review to approved. Way to go!

File: cypress/support/commands.js
Comment (davidmedina-contact) [inline cypress/support/commands.js:1]: You can clean this up to remove default Cypress comments

File: cypress/support/e2e.js
Comment (davidmedina-contact) [inline cypress/support/e2e.js:1]: Same here

**Review by davidmedina-contact — CHANGES_REQUESTED** [review]
Good progress! The most imporant feedback is about making your tests more reliable by not relying on data that is externakl to Cypress that can change as you evolve your app and code

File: cypress/e2e/addProject.cy.js (hunk 10 above)
Comment (abbiehuynh) [reply cypress/e2e/addProject.cy.js]: I did it just to see it would pass without having to rewrite everything above twice, but I didn't think about it being a valuable functionality, which is a good point! 

I added a testing suite to test both routes and used the beforeEach to run the similar logic it takes to get to the individual routes. 

## Testing Screenshot:
add-project-modal-testing

Submitted changes with commit: [f26b6da](https://github.com/abbiehuynh/my-crochet-corner/pull/14/commits/f26b6daf1efdcdfe071d7fc3c6c3022103cc3791)

File: cypress/e2e/login.cy.js (hunk 16 above)
Comment (abbiehuynh) [reply cypress/e2e/login.cy.js:62]: Thank you, Tati!!

File: cypress/e2e/login.cy.js (hunk 14 above)
Comment (abbiehuynh) [reply cypress/e2e/login.cy.js]: Submitted changes with commit: [6206a20](https://github.com/abbiehuynh/my-crochet-corner/pull/14/commits/6206a20e146d09240c503101846f3c1212574095)

File: cypress/e2e/ai.cy.js (hunk 12 above)
Comment (abbiehuynh) [reply cypress/e2e/ai.cy.js:35]: This was also really fascinating for me to watch through Cypress's live testing suite!! It's so nice to be able to share this excitement!

File: client/src/components/auth/Login.jsx (hunk 20 above)
Comment (abbiehuynh) [reply client/src/components/auth/Login.jsx:51]: Line 48 is a useState created for error (error, setError) to display a message to the user below the login form, while line 49 is a console logged error. 

In the component, the error is displayed below the form
Screenshot 2024-11-05 at 3 30 17 PM


And in the UI it can be shown here.
Screenshot 2024-11-05 at 3 27 02 PM

File: cypress/e2e/projectCard.cy.js (hunk 24 above)
Comment (abbiehuynh) [reply cypress/e2e/projectCard.cy.js:3]: it.only will only run the test that has it.only instead of running all the tests within the test file. Cypress automatically re-runs the test with any changes including typing a "space" so I used it.only on something I was okay re-rerunning until i was ready to run the entire test file. 
I then kept it there in case I accidentally open the spec file and did not want to delete exisiting projects

File: client/src/components/home/ProjectProvider.jsx (hunk 2 above)
Comment (abbiehuynh) [reply client/src/components/home/ProjectProvider.jsx:83]: Through testing, I realized it wasn't working correctly, so I updated it and then ran the tests for it

File: cypress.config.js (hunk 3 above)
Comment (abbiehuynh) [reply cypress.config.js]: Submitted changes with commit: [b1b7b4a](https://github.com/abbiehuynh/my-crochet-corner/pull/14/commits/b1b7b4ac0bf07f6bf40209ab18961910dd70c154)

File: cypress/e2e/addProject.cy.js (hunk 4 above)
Comment (abbiehuynh) [reply cypress/e2e/addProject.cy.js]: It is part of the project requirement to test rendering of all components, so I tested the presence of the button before testing user interaction with the button

File: cypress/e2e/addProject.cy.js (hunk 5 above)
Comment (abbiehuynh) [reply cypress/e2e/addProject.cy.js]: Submitted changes with commit: [fee56c3](https://github.com/abbiehuynh/my-crochet-corner/pull/14/commits/fee56c3e708a7ff01083358f045ae5f93970aa82)

File: cypress/e2e/addProject.cy.js (hunk 6 above)
Comment (abbiehuynh) [reply cypress/e2e/addProject.cy.js]: This was mentioned in a previous comment and been updated in commit: [f26b6da](https://github.com/abbiehuynh/my-crochet-corner/pull/14/commits/f26b6daf1efdcdfe071d7fc3c6c3022103cc3791)

File: cypress/e2e/ai.cy.js (hunk 8 above)
Comment (abbiehuynh) [reply cypress/e2e/ai.cy.js:7]: I am just using this specific user as the testing account

File: cypress/e2e/addProject.cy.js (hunk 7 above)
Comment (abbiehuynh) [reply cypress/e2e/addProject.cy.js]: Submitted changes with commit: [5b7040e](https://github.com/abbiehuynh/my-crochet-corner/pull/14/commits/5b7040e3e40f306da46a71f769af9e8db802c6f1)

File: cypress/e2e/ai.cy.js (hunk 9 above)
Comment (abbiehuynh) [reply cypress/e2e/ai.cy.js:34]: The AI response. 
I tried to implement that the test verifies the response based on receiving it instead of doing a timeout, but the test will automatically fail within 4-5 seconds and the AI needs more time to load a response. It would not wait even when i tried cy.wait for the response instead of timeout.

File: cypress/e2e/home.cy.js (hunk 11 above)
Comment (abbiehuynh) [reply cypress/e2e/home.cy.js:37]: It was part of the project requirement to make sure everything rendered on the webpage.

File: cypress/e2e/navbar.cy.js (hunk 15 above)
Comment (abbiehuynh) [reply cypress/e2e/navbar.cy.js:6]: This is very interesting and informative! I will definitely look into implementing this change as my test and app grows, especially in testing my app 2.0!

File: cypress/e2e/projectCard.cy.js (hunk 17 above)
Comment (abbiehuynh) [reply cypress/e2e/projectCard.cy.js]: Ah, I think that this was just something that was missed when implementing CSS changes. I will make note to update that during a clean up PR

File: cypress/e2e/projectCard.cy.js (hunk 18 above)
Comment (abbiehuynh) [reply cypress/e2e/projectCard.cy.js:20]: Submitted changes with commit: [edb2e10](https://github.com/abbiehuynh/my-crochet-corner/pull/14/commits/edb2e10808b225a45eedd203539c15a13207a734)

File: cypress/e2e/projectCard.cy.js (hunk 19 above)
Comment (abbiehuynh) [reply cypress/e2e/projectCard.cy.js]: Changed it to compare to the (initial length - 1) of the project cards
Submitted change with commit: [edb2e10](https://github.com/abbiehuynh/my-crochet-corner/pull/14/commits/edb2e10808b225a45eedd203539c15a13207a734)

File: cypress/e2e/searchBar.cy.js (hunk 22 above)
Comment (abbiehuynh) [reply cypress/e2e/searchBar.cy.js:34]: Yes, I have these projects in the db for test purposes. I think something I can implement in the future is providing every new user with 3 template project cards (one for each project status) and create tests off of those?

File: cypress/e2e/sortProjectStatus.cy.js (hunk 23 above)
Comment (abbiehuynh) [reply cypress/e2e/sortProjectStatus.cy.js:38]: Something I want to implement in the future is a drop down so the user must choose from these options instead of being able to type it in!

File: cypress/fixtures/example.json
Comment (abbiehuynh) [reply cypress/fixtures/example.json:1]: Submiited changes with commit: [5b2e7f3](https://github.com/abbiehuynh/my-crochet-corner/pull/14/commits/5b2e7f39da62fa55315352ef8ae0c7f5f91a2f6f)

File: cypress/support/commands.js
Comment (abbiehuynh) [reply cypress/support/commands.js:1]: Submitted changes with commit: [c6fb501](https://github.com/abbiehuynh/my-crochet-corner/pull/14/commits/c6fb50185c579dbd97defaa09b653b5f25a295bb)

File: cypress/support/e2e.js
Comment (abbiehuynh) [reply cypress/support/e2e.js:1]: Submitted changes with commit: [b06e6c3](https://github.com/abbiehuynh/my-crochet-corner/pull/14/commits/b06e6c383de8cb2d6ecfc3286413c86871b6ab6e)

File: cypress/e2e/login.cy.js (hunk 13 above)
Comment (abbiehuynh) [reply cypress/e2e/login.cy.js:32]: This gives me a greater perspective on how to approach tests. I will remember to think about this when writing tests in future applications. 
For this final project, we are required to test all the components, even inputs in forms(even if its not something that we coded).

**Review by davidmedina-contact — APPROVED** [review]
LGTM
